data/provider/src/test/java/tech/pegasys/artemis/api/ValidatorDataProviderTest.java

```diff
@@ -102,6 +133,170 @@ private void shouldThrowDataProviderExceptionAfterGettingException(Exception ex)
+        new ValidatorDutiesRequest(
+            compute_epoch_at_slot(beaconState.slot), List.of(BLSPubKey.empty()));
+    when(validatorCoordinator.getProposerForSlot(any(), any())).thenReturn(pubKey1);
+    when(combinedChainDataClient.isStoreAvailable()).thenReturn(true);
+    when(combinedChainDataClient.getBestBlockRoot()).thenReturn(Optional.of(blockRoot));
+    when(combinedChainDataClient.getStateAtSlot(any(), any()))
+        .thenReturn(completedFuture(Optional.of(beaconStateInternal)));
+    when(combinedChainDataClient.getCommitteesFromState(any(), eq(beaconState.slot)))
+        .thenReturn(List.of());
+
+    SafeFuture<List<ValidatorDuties>> future = provider.getValidatorDutiesByRequest(smallRequest);
+    List<ValidatorDuties> validatorDuties = future.get();
+
+    assertThat(validatorDuties.size()).isEqualTo(1);
+    ValidatorDuties expected = new ValidatorDuties(BLSPubKey.empty(), null, null, List.of());
+    assertThat(validatorDuties.get(0)).isEqualToComparingFieldByField(expected);
+  }
+
+  @Test
+  void getValidatorDutiesByRequest_shouldIncludeValidatorDuties()
+      throws ExecutionException, InterruptedException {
+    // add a validator with a different pubkey since by default they are all the same
+    tech.pegasys.artemis.datastructures.state.BeaconState alteredInternalState =
+        addValidator(beaconStateInternal);
+
+    BeaconState alteredState = new BeaconState(alteredInternalState);
+    int addedValidatorIndex = alteredState.validators.size() - 1;
+    when(validatorCoordinator.getProposerForSlot(any(), any()))
+        .thenReturn(alteredInternalState.getValidators().get(addedValidatorIndex).getPubkey());
+
+    ValidatorDataProvider provider =
+        new ValidatorDataProvider(validatorCoordinator, combinedChainDataClient);
+    ValidatorDutiesRequest validatorDutiesByRequest =
+        new ValidatorDutiesRequest(
+            compute_epoch_at_slot(beaconState.slot),
+            List.of(
+                alteredState.validators.get(0).pubkey,
+                alteredState.validators.get(11).pubkey,
+                alteredState.validators.get(addedValidatorIndex).pubkey));
+    CommitteeAssignment ca1 = new CommitteeAssignment(List.of(3, 2, 1, 0), ZERO, alteredState.slot);
+    CommitteeAssignment ca2 =
+        new CommitteeAssignment(List.of(11, 22, 33, addedValidatorIndex), ZERO, alteredState.slot);
+    List<CommitteeAssignment> committeeAssignments = List.of(ca1, ca2);
+    List<UnsignedLong> allEpochSlots = getEpochSlotsFromStartSlot(beaconState.slot);
+
+    when(combinedChainDataClient.isStoreAvailable()).thenReturn(true);
+    when(combinedChainDataClient.getBestBlockRoot()).thenReturn(Optional.of(blockRoot));
+    when(combinedChainDataClient.getStateAtSlot(any(), any()))
+        .thenReturn(completedFuture(Optional.of(alteredInternalState)));
+    when(combinedChainDataClient.getCommitteesFromState(any(), eq(alteredInternalState.getSlot())))
+        .thenReturn(committeeAssignments);
```

Comment:
to have this much code before the function under test is of concern to me. i'm not sure if we need to refactor, break out the setup, or something, but this is huge for a unit test...